examples/state_machine/states.py

```diff
@@ -0,0 +1,471 @@
+            state=HandshakingState,
+            context=self._context
+        )
+
+    def subscription_restored(self, event: events.SubscriptionRestoredEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        return PNTransition(
+            state=ReceivingState,
+            context=self._context
+        )
+
+
+class HandshakeStoppedState(PNState):
+    def __init__(self, context: PNContext) -> None:
+        super().__init__(context)
+        self._transitions = {
+            events.ReconnectEvent.__name__: self.reconnect
+        }
+
+    def reconnect(self, event: events.ReconnectEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        return PNTransition(
+            state=HandshakeReconnectingState,
+            context=self._context
+        )
+
+
+class ReceivingState(PNState):
+    def __init__(self, context: PNContext) -> None:
+        super().__init__(context)
+        self._transitions = {
+            events.SubscriptionChangedEvent.__name__: self.subscription_changed,
+            events.SubscriptionRestoredEvent.__name__: self.subscription_restored,
+            events.ReceiveSuccessEvent.__name__: self.receiving_success,
+            events.ReceiveFailureEvent.__name__: self.receiving_failure,
+            events.DisconnectEvent.__name__: self.disconnect,
+            events.ReconnectEvent.__name__: self.reconnect,
+        }
+
+    def on_enter(self, context: Union[None, PNContext]):
+        super().on_enter(context)
+        return effects.ReceiveEventsEffect(context.channels, context.groups)
+
+    def on_exit(self):
+        super().on_exit()
+        return effects.CancelReceiveEventsEffect()
+
+    def subscription_changed(self, event: events.SubscriptionChangedEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        return PNTransition(
+            state=self.__class__, context=context,
+            context=self._context
+        )
+
+    def subscription_restored(self, event: events.SubscriptionRestoredEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        return PNTransition(
+            state=self.__class__, context=context,
+            context=self._context
+        )
+
+    def receiving_success(self, event: events.ReceiveSuccessEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        return PNTransition(
+            state=self.__class__, context=context,
+            context=self._context
+        )
+
+    def receiving_failure(self, event: events.ReceiveFailureEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        return PNTransition(
+            state=ReceiveReconnectingState,
+            context=self._context
+        )
+
+    def disconnect(self, event: events.DisconnectEvent, context: PNContext) -> PNTransition:
+        return PNTransition(state=ReceiveStoppedState)
```

Comment:
I think you need to pass here context as well, so the channels, channelGroups, tt and region will be preserved